# Vector Basics for Data Science

This project introduces fundamental vector operations using Python and NumPy. 

It explores:
- vector creation
- vector addition
- dot products
- vector magnitude
- cosine similarity

The notebook also demonstrates how vectors can be used a simplified recommendation-system example by comparing user preference vectors with item feature vectors.


# Using NumPy to Create Vectors as Arrays

Vectors can be created in Python using NumPy arrays. Each number entered into the array represents one feature or dimension of the vector. 

The number of inputs equals the number of dimensions the vector has. For our purposes, vectors must contain the same number of dimensions.

The following code creates two vectors that will be shown in the other examples throughout this notebook.

In [2]:
import numpy as np

v1 = np.array([1, 2, 3])
v2 = np.array([4, 5, 6])

v1, v2

(array([1, 2, 3]), array([4, 5, 6]))

## Experiment

Throughout this notebook, we will be building a movie recommendation system based on a user's genre preferences.

The five genres used in this project are:

1. Action
2. Comedy
3. Drama
4. Horror
5. Romance

Each vector will use a 0–5 rating scale, where:

- 0 = no preference or presence
- 1 = very low preference or presence
- 5 = very high preference or presence

In the `user_preferences` vector below, put in a comma-separated list of your 0–5 rating of how much you enjoy each genre respectively: action, comedy, drama, horror, romance.

Thus, the `user_preferences` vector represents how much the user enjoys each genre respectively.

Example: A user that prefers action, horror, and romance, but only slightly enjoys comedy and doesn't enjoy drama, should put:

```python
user_preferences = np.array([5, 2, 0, 4, 5])
```
Now input your ratings for each genre, and then for movies A and B, input the presence of each of these genres. They will be inputted into vectors and used for future reference in the project.

In [ ]:
import numpy as np

genres = ["Action", "Comedy", "Drama", "Horror", "Romance"]

def get_valid_rating(prompt):
    while True:
        try:
            rating = int(input(prompt))

            if rating in [1, 2, 3, 4, 5]:
                return rating
            else:
                print("Enter valid number\n")

        except ValueError:
            print("Enter valid number\n")


print("\nRate each genre from 1 to 5 based on how much you like it.\n")

user_ratings = []

for genre in genres:
    rating = get_valid_rating(f"{genre}: ")
    user_ratings.append(rating)

user_preferences = np.array(user_ratings)

print("\nUser preference vector:\n")
print(user_preferences)


print("\nEnter genre ratings for Movie A from 1 to 5. Remember the genres are Action, Comedy, Drama, Horror, Romance respectively.\n")

movie_a_ratings = []

for genre in genres:
    rating = get_valid_rating(f"Movie A - {genre}: ")
    movie_a_ratings.append(rating)

movie_a = np.array(movie_a_ratings)


print("\nEnter genre ratings for Movie B from 1 to 5. Remember the genres are Action, Comedy, Drama, Horror, Romance respectively.\n")

movie_b_ratings = []

for genre in genres:
    rating = get_valid_rating(f"Movie B - {genre}: ")
    movie_b_ratings.append(rating)

movie_b = np.array(movie_b_ratings)

print("\nMovie A vector:\n")
print(movie_a)

print("\nMovie B vector:\n")
print(movie_b)

These movie ratings will be saved as we move through this project.

# Vector Addition

Vector addition adds the matching dimensions in each vector.

In data science, vector addition depicts accumulating features, signals, or measurements of distinct data points.

The code below shows how to add two vectors.

In [3]:
v_add = v1 + v2
v_add


array([5, 7, 9])

## Result Interpretation

The result is `[5, 7, 9]` because NumPy adds each matching position:

- 1 + 4 = 5
- 2 + 5 = 7
- 3 + 6 = 9

## Experiment

Now we can apply vector addition to the movie vectors we created earlier.

When we add two movie vectors together, we create a combined genre profile. This shows the total genre strength across both movies.

Run the cell below to see the combined genre profile of your two movies.


In [ ]:
combined_movie_profile = movie_a + movie_b

print("\nCombined movie profile:")
print(combined_movie_profile)
print()

for genre, value in zip(genres, combined_movie_profile):
    print(f"{genre}: {value}")

The output shows how strongly each genre is represented after adding the Movie A and Movie B vectors. Higher numbers mean that genre is more strongly represented across the two movies than the others

# Dot Product

The dot product is a mathematical operation that combines two vectors into a single number. This number helps measure the relationship between the vectors.

It works by:

1. multiplying the matching values in each vector
2. adding all of those products together

In data science, the dot product is important because it helps measure how strongly two vectors align with each other.

In this project, the dot product is used to compare the user preference vector with the `movie_a` or `movie_b` vector. When a dot product is taken between the preference vector and a movie vector, the program uses this operation to measure how strongly the two vectors align with each other.

The result of the dot product then becomes part of the cosine similarity calculation, which will be discussed later.

In [4]:
dot_product = np.dot(v1, v2)

print(dot_product)

32


## Result Interpretation

The result is `32` because NumPy multiplies each matching position, then adds the products together:

- 1 × 4 = 4
- 2 × 5 = 10
- 3 × 6 = 18

Then:

- 4 + 10 + 18 = 32

So, the dot product of `[1, 2, 3]` and `[4, 5, 6]` is `32`.

## Experiment

Now, we can use the dot product to compare the user's preference vector with each movie vector.

A higher dot product suggests the movie's genre profile aligns more strongly with the user's preferences.

In [ ]:
user_movie_a_alignment = np.dot(user_preferences, movie_a)
user_movie_b_alignment = np.dot(user_preferences, movie_b)

print("\nUser-Movie A dot product:", user_movie_a_alignment)
print("\nUser-Movie B dot product:", user_movie_b_alignment)

if user_movie_a_alignment > user_movie_b_alignment:
    print("\nMovie A is a better match based on the dot product.")
elif user_movie_b_alignment > user_movie_a_alignment:
    print("\nMovie B is a better match based on the dot product.")
else:
    print("\nBoth movies match the user equally based on the dot product.")

The dot product helps compare the user's match with each movie by measuring how strongly the user's preference vector aligns with each movie vector. The movie with the higher dot product is the stronger match based on this metric.

However, the dot product does not fully account for scale. A higher number may mean stronger alignment, but it may also mean that one vector simply has larger values overall. The next techniques, magnitude and cosine similarity, will help resolve this issue.



# Vector Magnitude

Vector magnitude is the length or size of a vector, regardless of its direction. The vector magnitude is essentially calculated using the Pythagorean theorem.

In data science, magnitude is important because it helps us understand the scale of a vector. Dot products can produce higher values simply because one or both vectors contain larger numbers overall.

A technique called cosine similarity uses magnitude to normalize vectors, which allows us to compare their similarity more precisely than just using the size of their dot products.

In the movie experiment, the movie vector with a larger magnitude may have larger genre values overall, but that does not automatically mean it is the best match for the user. This is why magnitude becomes important when we calculate cosine similarity.

In [5]:
magnitude = np.linalg.norm(v1)

print(round(magnitude, 2))

3.74


## Result Interpretation

The result is approximately `3.74` because the magnitude measures the length of the vector `v1`.

For `v1 = [1, 2, 3]`:

- 1² = 1
- 2² = 4
- 3² = 9

Add them together:

- 1 + 4 + 9 = 14

Take the square root:

- √14 ≈ 3.74

So the magnitude (length) of vector `v1` is approximately `3.74`.

## Experiment: Measuring Vector Magnitude

Now, we can calculate the magnitude of the user preference, Movie A, and Movie B vectors.

This helps us measure the overall size of these vectors.

Larger magnitudes mean the vector contains larger values overall, but that does not automatically mean it is a better recommendation. 

Run the cell below to see the magnitude of each vector.

In [ ]:
user_magnitude = np.linalg.norm(user_preferences)
movie_a_magnitude = np.linalg.norm(movie_a)
movie_b_magnitude = np.linalg.norm(movie_b)

print("User magnitude:", round(user_magnitude, 2))
print("Movie A magnitude:", round(movie_a_magnitude, 2))
print("Movie B magnitude:", round(movie_b_magnitude, 2))

NameError: name 'user_preferences' is not defined

The magnitudes show the overall size or scale of each vector.

A movie may have a larger magnitude because it has higher genre values overall, but this can artificially increase the dot product. This can still be an issue even when using a smaller 0–5 rating scale.

This is why the final technique, cosine similarity, becomes important. It reduces the effect of larger magnitudes, allowing us to compare the direction or pattern of the vectors more fairly.

# Cosine Similarity

Cosine similarity measures how similar two vectors are by comparing their direction rather than just their size.

This is important in data science because two vectors can have very different magnitudes but still follow a similar pattern. Cosine similarity helps adjust for scale, making the comparison more fair.

In this project, cosine similarity helps compare the user's preference vector with each movie vector. Instead of only asking which movie has the highest dot product, cosine similarity asks which movie has the most similar genre pattern to the user's preferences.

In [10]:
cos_sim = dot_product / (np.linalg.norm(v1) * np.linalg.norm(v2))

print(round(cos_sim,2))

0.97


## Result Interpretation

The result is approximately `0.97` because cosine similarity measures how similar the direction or pattern of two vectors is.

For:

`v1 = [1, 2, 3]`  
`v2 = [4, 5, 6]`

First, find the dot product:

- 1 × 4 = 4
- 2 × 5 = 10
- 3 × 6 = 18

Add them together:

- 4 + 10 + 18 = 32

Next, find the magnitude of each vector.

For `v1`:

- 1² + 2² + 3² = 14
- √14 ≈ 3.74

For `v2`:

- 4² + 5² + 6² = 77
- √77 ≈ 8.77

Now use the cosine similarity formula:

`cosine similarity = dot product / (magnitude of v1 × magnitude of v2)`

So:

- 32 / (3.74 × 8.77)
- 32 / 32.80
- ≈ 0.97

So the cosine similarity between `v1` and `v2` is approximately `0.97`.

Because this value is close to `1`, it means the two vectors point in a very similar direction. 

## Experiment: Using Cosine Similarity for Movie Recommendations

Now, we can use cosine similarity to compare the user's preferences with each movie vector.

The closer the cosine similarity score is to 1, the more similar the movie vector is to the user's preference vector.

In [ ]:
def cosine_similarity(a, b):
    a_magnitude = np.linalg.norm(a)
    b_magnitude = np.linalg.norm(b)

    if a_magnitude == 0 or b_magnitude == 0:
        print("Cosine similarity cannot be calculated when one vector has a magnitude of 0.")
        return 0
    
    return np.dot(a, b) / (a_magnitude * b_magnitude)


movie_a_cos_sim = cosine_similarity(user_preferences, movie_a)
movie_b_cos_sim = cosine_similarity(user_preferences, movie_b)

print("Movie A cosine similarity:", round(movie_a_cos_sim, 3))
print("Movie B cosine similarity:", round(movie_b_cos_sim, 3))

if movie_a_cos_sim > movie_b_cos_sim:
    print("Recommended: Movie A")
elif movie_b_cos_sim > movie_a_cos_sim:
    print("Recommended: Movie B")
else:
    print("Both movies are equally similar.")

Cosine similarity gives a scale-adjusted similarity score.

Unlike the dot product, cosine similarity accounts for vector magnitude, so it is better for comparing the pattern of the user's preferences against each movie's genre profile.

This makes the recommendation more fair when one movie vector has larger values overall than another.

# Conclusion

In this project, we used vectors to represent user preferences and movie genre profiles.

We practiced several important vector operations:

- vector creation
- vector addition
- dot product
- vector magnitude
- cosine similarity

The dot product helped compare how strongly the user preference vector aligned with each movie vector, but it could be affected by larger overall values.

Cosine similarity improved the comparison by accounting for vector magnitude and focusing more on the actual pattern or direction of the vectors.

Based on cosine similarity, the program recommends the movie with the genre pattern most similar to the user's preferences. Run the cell below to see your reccomended movie:

In [ ]:
print("\nFinal Recommendation:")

if movie_a_cos_sim > movie_b_cos_sim:
    print("Movie A is the best recommendation based on cosine similarity.")
elif movie_b_cos_sim > movie_a_cos_sim:
    print("Movie B is the best recommendation based on cosine similarity.")
else:
    print("Both movies are equally good recommendations based on cosine similarity.")